helpers/oauth/oauth.go

```diff
@@ -44,34 +44,59 @@ var client = &http.Client{
 //   token               - Auth token being used for the request
 //   repo                - Repo you are requesting access too e.g. bainsy88/busybox
 //   username            - Username for the OAuth request, identifies the type of token being passed in. Valid usernames are token (for registry token), iambearer, iamapikey, bearer (UAA bearer (legacy)), iamrefresh
-//   writeAccessRequired - Whether or not you require write (push and delete) access as well as read (pull)
 //   service             - The service you are retrieving the OAuth token for. Current services are either "notary" or "registry"
 //   hostname            - Hostname of the registry you wish to call e.g. https://registry.ng.bluemix.net
 // Returns:
 //   *auth.TokenResponse - Details of the type is here https://github.ibm.com/alchemy-registry/registry-types/tree/master/auth#type-tokenresponse
 //                         Token is the element you will need to forward to the registry/notary as part of a Bearer Authorization Header
 //   error
-func Request(token string, repo string, username string, writeAccessRequired bool, service string, hostname string) (*TokenResponse, error) {
-	var actions string
-	//If you want to verify if a the credential supplied has read and write access to the repo we ask oauth for pull,push and *
-	if writeAccessRequired {
-		actions = "pull,push,*"
-	} else {
-		actions = "pull"
+func Request(token string, repo string, username string, challengeSlice []Challenge) (*TokenResponse, error) {
+
+	oauthEndpoint := ""
+	service := ""
+	scope := ""
+
+	if challengeSlice == nil {
+		errMessage := "unable to fetch www-authenticate header details"
+		glog.Errorf(errMessage)
+		return nil, fmt.Errorf(errMessage)
 	}
 
-	resp, err := client.PostForm(hostname+"/oauth/token", url.Values{
-		"service":    {service},
-		"grant_type": {"password"},
-		"client_id":  {"testclient"},
-		"username":   {username},
-		"password":   {token},
-		"scope":      {"repository:" + repo + ":" + actions},
-	})
+	for _, challenge := range challengeSlice {
+		oauthEndpoint = challenge.Parameters["realm"]
+		service = challenge.Parameters["service"]
+		scope = challenge.Parameters["scope"]
+	}
```

Comment:
If we had multiple challenges in the list, we'd end up with the realm, service, and scope from the last one. I think we'd only ever have one Challenge though, so I think that's okay... is that right?